### ISSUE TITLE
Unstopped threads with net.logstash.logback.appender.LogstashTcpSocketAppender

### ISSUE THREAD

**Issue by serphacker** [logstash/logstash-logback-encoder#74]
Everything is working fine but some threads aren't stopped when the main return. How to stop them ?

```java
public class App {
    public final static Logger LOG = LoggerFactory.getLogger(App.class);
    public static void main(String[] args) {
        LOG.info("msg x2");
    }
}
```
```xml
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    
    <appender name="stash" class="net.logstash.logback.appender.LogstashTcpSocketAppender">
        <remoteHost>127.0.0.1</remoteHost>
        <port>1324</port>
        <encoder class="net.logstash.logback.encoder.LogstashEncoder" />
    </appender>
    <root level="all">
        <appender-ref ref="stash" />
    </root>
</configuration>
```

**Comment by philsttr** [comment]
Hi @serphacker,

Currently, the writer thread that is created by `LogstashTcpSocketAppender` inherits the daemon status from the thread that spawned it.  Since the main thread is a non-daemon thread, it means that the worker thread will be a non-daemon thread.  Non-daemon threads prevent the JVM from shutting down if they are still running.

This is a bug, and I will change the behavior to make the worker threads be daemon threads, so that they do not prevent the JVM from shutting down.

In the meantime, you can tell logback to shut down as follows:

```
import org.slf4j.ILoggerFactory;
import org.slf4j.LoggerFactory;

import ch.qos.logback.classic.LoggerContext;

        ILoggerFactory factory = LoggerFactory.getILoggerFactory();
        if (factory instanceof LoggerContext) {
            LoggerContext ctx = (LoggerContext)factory;
            ctx.stop();
        }
```

You can do this from the end of your main method, or add a JVM shutdown hook (`Runtime.getRuntime().addShutdownHook`) that does it.

**Comment by philsttr** [comment]
Fixed for version 4.1 (not released yet)

**Comment by evpaassen** [comment]
This bug is bothering me too. The workaround by adding a shutdown hook won't work, as the JVM will not be shutdown as long there is still a user thread (unless an exception raises). I'm looking forward to the 4.1 release.
